### PR TITLE
feat(entities): add draft mutateEntities

### PR DESCRIPTION
**Pull request by zubko**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngneat/elf/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

It's impossible to conditionally choose a mutation operation based on some property of mutated entity

Continuation of:
* Discussion: https://github.com/ngneat/elf/discussions/279
* PR: https://github.com/ngneat/elf/pull/284

## What is the new behavior?

It's possible to conditionally choose a mutation operation based on some property of mutated entity

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

### CODE REVIEW

**Comment by zubko** [comment]
@NetanelBasal I've made a new PR, because this way is very different from the original to keep it in the same PR.

I will move it to another file, I'm just temporarily highjacking the existing infra of update :) 

Some points not very clear for me yet:

* How to name the file? `mutate.mutation.ts` ? :) (I understand now why the word `mutate` can work here, but it will be funny to follow the naming pattern, maybe just `mutate.ts` then, since this is some kind of abstract "mutation frame" for the real mutators :) )

* The API of the callback 🤔 - the bad thing that sometimes we won't have an Entity (for example if we want to make some kind of `upsert`), then in the callback we will have to put the checks even if we know that the Entity is there for sure, but just because the API forces us, I'm not sure if this can be avoided considering the general purpose of the function. But anyways there is question of how to structure that API, as separate args, as an Array or Object etc.

**Comment by NetanelBasal** [comment]
It's a good question. I don't think it's flexible enough. I think we need to think of more options.

**Comment by NetanelBasal** [comment]
Can you summarize where it all began?

**Comment by zubko** [comment]
I have a reducer in my code which is doing an opposite to `upsert`, (updelete) , so I was suggesting to add an opposite to `upsert` to the API of Elf :)

**Comment by NetanelBasal** [comment]
Which way do you currently use to do it?

**Comment by zubko** [comment]
I have a super specialized reducer for the start:
```ts
export const updateOrRemoveProductIdFromCart = (productId: Product["id"]) => {
  store.update((state) => {
    const entity = state.entities[productId];
    if (!entity) {
      return state;
    }
    if (entity.count > 1) {
      const newEntity: typeof entity = { ...entity, count: entity.count - 1 };
      return {
        ...state,
        entities: { ...state.entities, [entity.productId]: newEntity },
      };
    } else {
      const newEntities = { ...state.entities };
      Reflect.deleteProperty(newEntities, entity.productId);
      const newIds = [...state.ids];
      newIds.splice(newIds.indexOf(entity.productId), 1);
      return { ...state, entities: newEntities, ids: newIds };
    }
  });
}
``` 

I was planning either to extract the behavior of `up-delete` and use it like:
```ts
  store.update(
    updateOrDelete(id, {
      deleteCondition: (item) => item.count <= 1,
      update: ({ count }) => ({ count: count - 1 }),
    })
  );
```

Or simply getting the entity first, since this is just one entity in the very small store:
```ts
  const cartItem = store.query(getEntity(productId));
  if (!cartItem) {
    return;
  }
  store.update(
    cartItem.count > 1
      ? updateEntities(productId, { count: cartItem.count - 1 })
      : deleteEntities(productId)
  );
```

I like extracting the behavior, then the code will require less deduction to understand what is going on. (Of course my original version is a total disaster is the sense of mixing the levels of details)

**Comment by NetanelBasal** [comment]
The `updateOrDelete` `update` option should support the same options as `updateEntities`

**Comment by NetanelBasal** [comment]
That's why I think the second option isn't so bad. The question is how common is this use case.

**Comment by NetanelBasal** [comment]
We can try the following:

```ts
  store.update(
    predicateUpdate(
      getEntity(1)?.count > 1,
      updateEntities(productId, { count: cartItem.count - 1 }),
      deleteEntities(productId)
    )
  );
```

**Comment by zubko** [comment]
> The `updateOrDelete` `update` option should support the same options as `updateEntities`

Do you mean that it should support another ref, like changing UI entities?

> That's why I think the second option isn't so bad. The question is how common is this use case.

Not super common, that's for sure. But in my opinion if there is `updateOrInsert` so `updateOrDelete` is kind-of symmetrical. And symmetry is good :) (in generic APIs). But as there are people as there are opinions, so no problem if you don't think it's worth adding it. From my perspective it's more like "yes-yes-no-no", if you know what I mean - the issue is so small that too much time discussing it would not make much sense.

Anyways, it's great that the API is made is in such a way so that a custom mutators / queries etc can be added and they will look like 1st class citizens as if they are part of core API. As I understand it came from the wish for tree-shaking, but it also helps here - I can have any crazy Entity APIs I want on my project and they will look the same as the common API.

> We can try the following:
> 
> ```ts
>   store.update(
>     byPredicate(
>       getEntity(1)?.count > 1,
>       updateEntities(productId, { count: cartItem.count - 1 }),
>       deleteEntities(productId)
>     )
>   );
> ```

Is that very different from the one in this PR?

```ts
store.update(
        mutateEntities([1, 2], (id, todo) =>
          todo?.completed
            ? deleteEntities(id)
            : updateEntities(id, { completed: true })
        )
      );
```

I think `mutateEntities` allows `if/else if/else` conditions while `byPredicte` is just `if/else` (or you'd have to nest it). And downside of using optionals is the same in both variants. Also where does the `getEntity` comes from?

**Comment by NetanelBasal** [comment]
I was thinking about the use case of addEntities. In both cases, it's an issue. Let's see if we get an additional request for this feature. For now, it's too "complicated," and I don't think it's worth it. 

I appreciate your contributions. Keep it coming!
